shellwhat/test_exercise.py

```diff
@@ -30,10 +30,11 @@ def test_exercise(
         solution_conn=solution_conn,
         student_result=student_result,
         solution_result=solution_result,
-        reporter=Reporter(error),
+        reporter=Reporter(errors=error),
         force_diagnose=force_diagnose,
     )
 
+    State.root_state = state
```

Comment:
Is it necessary to use State as a globalton? (maybe being done for backwards compat?). It seems like since you define `SCT_CTX`, even if done for backwards compat with old funcs (e.g. `success_msg`), you should be able to define them in `SCT_CTX`, such that they have access to the current state instance, without needing to pull it via modifying an attribute in a class in a module (State.root_state).

(could be very wrong here)

Reply:
`report()` highlights the current focus, but not if it's all the code.
`root_state` is used to know if current == all:
https://github.com/datacamp/protowhat/blob/7b6002d32c5104a31de3e43c06b994aac6b60d19/protowhat/State.py#L129

Reply:
With the new state linking mechanism in protowhat, one alternative to a class property would be to do this:
```
    @property
    def root_state(self):
        return self.get_state_history()[0]
```